File: dotnet/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example2/Handlers/ContentValidatorHandler.cs
namespace ChainOfResponsibilityDesignPattern.Example2.Handlers;

public class ContentValidatorHandler : BaseFileHandler
{
    private readonly List<string> _forbiddenWords = new() 
    { 
        "CONFIDENTIAL_LEAK", "SECRET_DATA", "BANNED_CONTENT" 
    };

    public override void Handle(FileUploadRequest request)
    {
        // Check file content for forbidden words
        string contentAsString = System.Text.Encoding.UTF8.GetString(request.FileContent);
        
        var foundForbiddenWords = _forbiddenWords
            .Where(word => contentAsString.Contains(word, StringComparison.OrdinalIgnoreCase))
            .ToList();

        if (foundForbiddenWords.Any())
        {
            request.IsValid = false;
            request.ValidationMessages.Add($"Forbidden content detected: {string.Join(", ", foundForbiddenWords)}");
            return; // Stop the chain
        }
        
        // Pass to next handler
        base.Handle(request);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example2/src/Handlers/ContentValidatorHandler.ts
import { FileUploadRequest } from "../FileUploadRequest";
import { BaseFileHandler } from "../BaseFileHandler";

const FORBIDDEN_WORDS = ["CONFIDENTIAL_LEAK", "SECRET_DATA", "BANNED_CONTENT"];

export class ContentValidatorHandler extends BaseFileHandler {
  handle(request: FileUploadRequest): void {
    const content = request.fileContent.toString("utf8");
    const found = FORBIDDEN_WORDS.filter((w) => content.toUpperCase().includes(w));
    if (found.length > 0) {
      request.isValid = false;
      request.validationMessages.push(
        `Forbidden content detected: ${found.join(", ")}`
      );
      return;
    }
    super.handle(request);
  }
}
